feat: generate endpoints using replica numbers

abeatrix opened the pull request:
RE:  https://github.com/sourcegraph/deploy-sourcegraph/pull/4213 & https://github.com/sourcegraph/sourcegraph/issues/45861

## Summary

**This feature is also backward-compatible, so existing setups and instances will not be affected by this change.**

This PR adds the ability to accept replicas number for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` via existing env vars (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`), and then generate a new list of endpoints based on replicas number instead of listing out each replica endpoint manually. 

### Background

Issue: Our current default setup only works in RBAC-enabled clusters. 

RBAC for frontend is used for service discovery as it is essential for frontend to arrange communications between services; however, this default setup often leads to delays and frustrating deployment experiences for those without the ability to create RBAC resources. As this is a common deployment pain point and issue that has negative impacts on both the customers and our business, our team believes this is something we should address. 

### Proposed Solution

A new base cluster introduces in https://github.com/sourcegraph/deploy-sourcegraph/pull/4215 provide an option for users to set up a working Sourcegraph k8s instance with or without RBAC enabled for frontend. 

When RBAC resources are removed from frontend, Sourcegraph will not work properly unless a list of services endpoints is included in the frontend env var list using existing env vars that are also being used in our[ docker-compose deployment ](https://github.com/sourcegraph/deploy-sourcegraph-docker/blob/master/docker-compose/docker-compose.yaml#L136-L147), which is a task that relies on users inputting the correct values manually.

The functions introduce in this PR aim to minimize the time our users/customers have to spend configuring their deployment (e.g. spending time inputting the endpoint addresses for each service manually) by generating the endpoints in frontend based on the replica number provided via env vars. 

**IMPORTANT** The changes made in this PR should **NOT** break any existing setup or instances. No additional changes are required in existing manifests.

### Expected Behavior

The current behavior is the same for **all** instances (both existing instances and new instances), which means:
- **ALL**  instances that do not have the **existing** `endpoints URLs` env vars set are default to using Kubernetes service discovery 
- Setting env vars to empty string "" will return ""
- See the Examples section listed below for the expected behaviors

#### Examples:

Return endpoints generated with the replica number:
```yaml
        env:
        - name: SRC_GIT_SERVERS
          value: "4"
        - name: SEARCHER_URL
          value: "4"
        - name: SYMBOLS_URL
          value: "4"
        - name: INDEXED_SEARCH_SERVERS
          value: "4"
```


## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

1. copy and paste the context below to your `sg.config.overwrite.yaml` file in the root of this directory:
2. run `sg start enterprise`
3. You should see frontend and gitserver complaining about the new endpoints that do not exist, with the newly generated list of endpoints in the correct format for k8s deployments.

The expected new endpoints using the file shared below should look like:
1. 4 gitserver endpoints: `"gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"`
2. 3 indexed search endpoints: `"indexed-search-0.indexed-search:6070", "indexed-search-1.indexed-search:6070", "indexed-search-2.indexed-search:6070"`
3. 3 symbols endpoints: `"http://symbols-0.symbols:3184", "http://symbols-1.symbols:3184", "http://symbols-2.symbols:3184"`
4. 3 searcher endpoints: `"http://searcher-0.searcher:3181", "http://searcher-1.searcher:3181", "http://searcher-2.searcher:3181"`

`sg.config.overwrite.yaml`:
```yaml
env:
  SRC_GIT_SERVERS: "5"
  SEARCHER_URL: "3"
  SYMBOLS_URL:"3"
  INDEXED_SEARCH_SERVERS: "3"
  DEPLOY_TYPE: kubernetes
```

Example output from running sg start:
```sh
[    gitserver-1] WARN server server/cleanup.go:193 current shard is not included in the list of known gitserver shards, will not delete repos {"current-hostname": "127.0.0.1:3502", "all-shards": ["gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"]}
[    gitserver-0] WARN server server/cleanup.go:193 current shard is not included in the list of known gitserver shards, will not delete repos {"current-hostname": "127.0.0.1:3501", "all-shards": ["gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"]}
[       frontend] ERROR InternalHandler httpapi/httpapi.go:281 API HTTP handler error response {"method": "POST", "request_uri": "/.internal/repos/index", "status_code": 500, "error": "hostname \"localhost:3070\" not found in Endpoints{\"indexed-search-0.indexed-search:6070\" \"indexed-search-1.indexed-search:6070\" \"indexed-search-2.indexed-search:6070\"}"}
[       frontend] ERROR internal.http trace/httptrace.go:250 unexpected status code 500 {"route_name": "internal.repos.index", "method": "POST", "url": "/.internal/repos/index", "code": 500, "duration": "2.073375ms", "shouldTrace": false, "error": "hostname \"localhost:3070\" not found in Endpoints{\"indexed-search-0.indexed-search:6070\" \"indexed-search-1.indexed-search:6070\" \"indexed-search-2.indexed-search:6070\"}"}
```

beyang commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -482,6 +485,13 @@ func gitservers() *endpoint.Map {
 	gitserversOnce.Do(func() {
 		gitserversVal = endpoint.New(func() string {
 			v := os.Getenv("SRC_GIT_SERVERS")
+			r := os.Getenv("GITSERVER_REPLICA_NUMBER")
```
```suggestion
			r := os.Getenv("GITSERVER_REPLICAS")
```

beyang commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
 - Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)
 - Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)
+- Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the correct environment variables (`GITSERVER_REPLICA_NUMBER`, `INDEXED_SEARCH_REPLICA_NUMBER`, `SEARCHER_REPLICA_NUMBER`, `SYMBOLS_REPLICA_NUMBER`) are provided in Kubernetes and Docker-Compose deployments
```
Rename these to `*_REPLICA_COUNT` instead of `*_REPLICA_NUMBER` to be more concise.

beyang commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -470,6 +471,8 @@ func (c *configurationSource) WriteWithOverride(ctx context.Context, input conft
 	return nil
 }
 
+var deploymentType = env.Get("DEPLOY_TYPE", "", "deployment type")
```
It feels a bit brittle to have the `DEPLOY_TYPE` env var toggle the base name used to refer to zoekt shards. Could we instead have the base name be configurable, as well? E.g., `GITSERVER_BASENAME` or `INDEXED_SEARCH_BASENAME` (default value `gitserver` and `indexed-search`)?

beyang commented on cmd/symbols/build-ctags.sh:
```diff
@@ -15,6 +15,7 @@ fi
 # Build ctags docker image for universal-ctags-dev
 echo "Building ctags docker image"
 docker build -f cmd/symbols/Dockerfile -t ctags . \
+  --platform linux/amd64 \
```
Why was this necessary?

abeatrix replied:
This enables non-amd64 systems to run amd64 images using docker. The original command does not work on m1 laptops (arm64), and adding [--platform linux/amd64](https://docs.docker.com/build/building/multi-platform/) will not break the command for amd64 systems 😄

beyang commented on internal/symbols/client.go:
```diff
@@ -76,12 +79,28 @@ type Client struct {
 			c.endpoint = endpoint.New(c.URL)
 		}
+		// Set endpoint as empty if both env vars are empty or invalid
+		if err != nil && len(strings.Fields(c.URL)) == 0 {
```
This seems like it should be part of the previous if-else block.

beyang commented on internal/symbols/client.go:
```diff
@@ -76,12 +79,28 @@ type Client struct {
-		if len(strings.Fields(c.URL)) == 0 {
-			c.endpoint = endpoint.Empty(errors.New("a symbols service has not been configured"))
+		// compute endpoints for symbols if replicas number is provided
+		num, err := strconv.Atoi(symbolsReplicas)
+		if err == nil && num > 0 {
```
```suggestion
		if num, err := strconv.Atoi(symbolsReplicas); err == nil && num > 0 {
```

beyang replied:
Can you factor this out into a unit test?

davejrt commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -479,19 +479,29 @@ var (
 
 func gitservers() *endpoint.Map {
+	const (
+		s = "gitserverh"
```
```suggestion
		s = "gitserver"
```

davejrt reviewed (commented):
Overall looks good. It be worth looking into the possibility of using existing envvars that are present in kuberetes or compose to determine the deployment type, rather than having to explicitly define it. For example checking for the presence of one of these could safely determine that it's k8s or not. 
```
/ $ env | grep KUBE
KUBERNETES_SERVICE_PORT=443
KUBERNETES_PORT=tcp://10.253.0.1:443
KUBERNETES_PORT_443_TCP_ADDR=10.253.0.1
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP=tcp://10.253.0.1:443
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_SERVICE_HOST=10.253.0.1
```

beyang commented:
It feels a bit hacky and toilsome to require setting both `SRC_${SERVICE}` and `${SERVICE}_REPLICA_COUNT`. What if we just errored on startup if both are set, and use whichever value is set?

beyang commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
 - Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)
 - Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)
+- Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL e ...
```
Endpoints for Gitserver, Indexed-search, Searcher, and Symbols are generated automatically when the replica counts for each service are provided using the correct environment variables (GITSERVER_REPLICA_COUNT, INDEXED_SEARCH_REPLICA_COUNT, SEARCHER_REPLICA_COUNT, SYMBOLS_REPLICA_COUNT) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (SEARCHER_URL, SYMBOLS_URL, INDEXED_SEARCH_SERVERS, SRC_GIT_SERVERS) must be set to kubernetes or docker-compose to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

beyang replied:
```suggestion
Endpoints for Gitserver, Indexed-search, Searcher, and Symbols are generated automatically when the replica counts for each service are provided using the correct environment variables (GITSERVER_REPLICA_COUNT, INDEXED_SEARCH_REPLICA_COUNT, SEARCHER_REPLICA_COUNT, SYMBOLS_REPLICA_COUNT) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (SEARCHER_URL, SYMBOLS_URL, INDEXED_SEARCH_SERVERS, SRC_GIT_SERVERS) must be set to kubernetes or docker-compose to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

beyang replied:
```suggestion
- Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service are provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

beyang commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
 - Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)
 - Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)
+- Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL e ...
```
I would suggest rephrasing to begin with mentioning the new environment variables, as this is the most relevant piece of information.

Added new environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) to the `sourcegraph-frontend` service to inform it of the replica count of sharded statefulsets. It uses these variables to determine the network address of each shard. These variables are intended to replace existing variables like `SRC_GIT_SERVERS`, which pass the shard addresses as a string, rather than relying on k8s naming conventions. If both `SRC_GIT_SERVERS` and `GITSERVER_REPLICA_COUNT` are set, `sourcegraph-frontend` will throw an error and fail to start.

abeatrix replied:
@beyang good point, thanks! What do you think about this?

```suggestion
- Added new environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) to the sourcegraph-frontend service to help it determine the network address of each shard in the sharded statefulsets based on replica counts. These new variables are intended to replace old ones like `SRC_GIT_SERVERS`. Sourcegraph will fail to start when both old and new variables are set. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

beyang commented on internal/endpoint/endpoint.go:
```diff
@@ -66,6 +68,67 @@ func New(urlspec string) *Map {
+// Note: Docker-compose and k8s deployments have different endpoints
+// docker-compose: zoekt-webserver-0:6070
+// k8s: indexed-search-0.indexed-search:6070
+func NewReplicas(urlspec string, service string, replicas string, port string, protocol string) *Map {
```
```suggestion
func NewReplicas(urlspec, service, replicas, port, protocol string) *Map {
```

rafax commented:
Two things I'd like to clarify as we're preparing to move Cloud from Docker Compose (v1.1) to Kubernetes (v2):

1. for Docker Compose, this feature is only enabled where customer opts-in by setting env variables, and the old approach  (SRC_GIT_SERVERS) will keep working for the foreseeable future (we [use it](https://sourcegraph.sourcegraph.com/search?q=context:global+SRC_GIT_SERVERS+r:%5Egithub%5C.com/sourcegraph/deploy-sourcegraph-managed%24+&patternType=standard&sm=1&groupBy=repo) in Cloud v1.1)?
1. for Kubernetes, if the newly introduced variables are not set we will default to using Kubernetes service discovery (i.e. merging this PR doesn't require changes in existing manifests)?

abeatrix commented:
@rafax apologize for not providing background info related to this PR! I've updated the PR but to answer you questions:

> for Docker Compose, this feature is only enabled where customer opts-in by setting env variables, and the old approach (SRC_GIT_SERVERS) will keep working for the foreseeable future (we [use it](https://sourcegraph.sourcegraph.com/search?q=context:global+SRC_GIT_SERVERS+r:%5Egithub%5C.com/sourcegraph/deploy-sourcegraph-managed%24+&patternType=standard&sm=1&groupBy=repo) in Cloud v1.1)?

The feature added is meant to enhance the current user experience and minimize the risks of human mistakes (eg. input the incorrect endpoints). Currently, it is designed to work with the existing envs. 
It is created with the intention of replacing the existing method of providing hostnames but we have yet to make plans to remove the existing environment variables as I know of.

> for Kubernetes, if the newly introduced variables are not set we will default to using Kubernetes service discovery (i.e. merging this PR doesn't require changes in existing manifests)?

This is correct. It is important to note that for the newly introduced variables to take effect, the existing environment variables must be updated accordingly as mentioned in the introduction above. 
There is [a PR](https://github.com/sourcegraph/sourcegraph/pull/46400) to replace the default k8s endpoint for symbols, but it is not related to anything I am working on.

I have also added more details to this[ PR intro above ](https://github.com/sourcegraph/sourcegraph/pull/45862#issue-1505250167) as well as the [PR that inspired the feature](https://github.com/sourcegraph/deploy-sourcegraph/pull/4215), which are both part of https://github.com/sourcegraph/pr-faqs/issues/9. Please feel free to let us know if you have other concerns or questions 😃

rafax commented:
> @rafax apologize for not providing background info related to this PR! I've updated the PR but to answer you questions:
> 
> > for Docker Compose, this feature is only enabled where customer opts-in by setting env variables, and the old approach (SRC_GIT_SERVERS) will keep working for the foreseeable future (we [use it](https://sourcegraph.sourcegraph.com/search?q=context:global+SRC_GIT_SERVERS+r:%5Egithub%5C.com/sourcegraph/deploy-sourcegraph-managed%24+&patternType=standard&sm=1&groupBy=repo) in Cloud v1.1)?
> 
> The feature added is meant to enhance the current user experience and minimize the risks of human mistakes (eg. input the incorrect endpoints). Currently, it is designed to work with the existing envs. It is created with the intention of replacing the existing method of providing hostnames but we have yet to make plans to remove the existing environment variables as I know of.

Understood, please let us (@sourcegraph/cloud) know if you plan to change the existing default / deprecate existing mechanism for Docker Compose (I hope we will be on Kubernetes by then). 

> > for Kubernetes, if the newly introduced variables are not set we will default to using Kubernetes service discovery (i.e. merging this PR doesn't require changes in existing manifests)?
> 
> This is correct. It is important to note that for the newly introduced variables to take effect, the existing environment variables must be updated accordingly as mentioned in the introduction above. There is [a PR](https://github.com/sourcegraph/sourcegraph/pull/46400) to replace the default k8s endpoint for symbols, but it is not related to anything I am working on.
> 
> I have also added more details to this[ PR intro above ](https://github.com/sourcegraph/sourcegraph/pull/45862#issue-1505250167) as well as the [PR that inspired the feature](https://github.com/sourcegraph/deploy-sourcegraph/pull/4215), which are both part of [sourcegraph/pr-faqs#9](https://github.com/sourcegraph/pr-faqs/issues/9). Please feel free to let us know if you have other concerns or questions 😃

My understanding here is that Cloud (running RBAC-enabled clusters) can keep using the current base cluster and built-in service discovery for the foreseeable future. Long-term, do you plan to migrate all customers using K8S to the new base cluster, or will you keep supporting both old and new base clusters?

beyang commented:
> know if you plan to change the existing default / deprecate existing mechanism for Docker Compose (I hope we will be on Kubernetes by then).

I doubt we'll stop supporting this anytime soon, as I think it'll take existing customers some time to fully migrate off Docker Compose.

> My understanding here is that Cloud (running RBAC-enabled clusters) can keep using the current base cluster and built-in service discovery for the foreseeable future.

This is doable, but would you need service discovery if the service names replica names follow strong conventions (e.g., `gitserver-0`, ..., `gitserver-n`)?

> Long-term, do you plan to migrate all customers using K8S to the new base cluster, or will you keep supporting both old and new base clusters - the [PRFAQ](https://docs.google.com/document/d/1pR29WO-D883QvhCQJlnOv_Qu5W59jf0ktWNhdYJd08c/edit#heading=h.f1qew9jn9vc2) mentions that you want to have 90% customers migrated to new base cluster (based on my understanding, no Helm & using changes from this PR) in 6 months, so we likely need to work on finding a way for Cloud to move in this direction (or we will have 10s or 100s of Kubernetes clusters still using Helm and old base cluster).

Long-term, we would like all kustomize customers to use the new base overlays. We will provide an overlay that reproduces the old base cluster from the new one to ease migration. We will push customers to adopt the new kustomize overlays, but we will likely need to keep Helm support around in some capacity. I'm curious what subset of the template Cloud is using—I'm guessing you're mainly interested in toggling replica counts and resource allocation? If that's the case, then I think the simplified Helm chart would continue to templatize those. To what extent would you need to do other stuff like apply custom security policies or change the namespace?

rafax commented:
> > know if you plan to change the existing default / deprecate existing mechanism for Docker Compose (I hope we will be on Kubernetes by then).
> 
> I doubt we'll stop supporting this anytime soon, as I think it'll take existing customers some time to fully migrate off Docker Compose.
Understood.
> > My understanding here is that Cloud (running RBAC-enabled clusters) can keep using the current base cluster and built-in service discovery for the foreseeable future.
> 
> This is doable, but would you need service discovery if the service names replica names follow strong conventions (e.g., `gitserver-0`, ..., `gitserver-n`)?

In general, ability to "just add replicas" would make Cloud operations easier (and would unblock use of HPA), but this already isn't possible for GitServer/Zoekt without modifying frontend. For now (while we are finalising and executing migration to v2) I'd want us to keep doing what we're doing (and have tested), going forward we can see how much horizontal scaling we end up doing, and it's possible that using a convention is fine.

> > Long-term, do you plan to migrate all customers using K8S to the new base cluster, or will you keep supporting both old and new base clusters - the [PRFAQ](https://docs.google.com/document/d/1pR29WO-D883QvhCQJlnOv_Qu5W59jf0ktWNhdYJd08c/edit#heading=h.f1qew9jn9vc2) mentions that you want to have 90% customers migrated to new base cluster (based on my understanding, no Helm & using changes from this PR) in 6 months, so we likely need to work on finding a way for Cloud to move in this direction (or we will have 10s or 100s of Kubernetes clusters still using Helm and old base cluster).
> 
> Long-term, we would like all kustomize customers to use the new base overlays. We will provide an overlay that reproduces the old base cluster from the new one to ease migration. We will push customers to adopt the new kustomize overlays, but we will likely need to keep Helm support around in some capacity. I'm curious what subset of the template Cloud is using—I'm guessing you're mainly interested in toggling replica counts and resource allocation? If that's the case, then I think the simplified Helm chart would continue to templatize those. To what extent would you need to do other stuff like apply custom security policies or change the namespace?

I'll defer to @michaellzc on specifics here, but from my understanding:
- Replica counts, resource allocation -> yes
- GCP specific config (like Workload Identity for Cloud SQL) 
- Cloud specific items (like Prometheus rules to drop high-cardinality metrics from centralized monitoring)
- we are working on adding NetworkPolicies to restrict pod-to-pod communication based on last Pentest
- over time, I'd like to move move towards "self-driving Cloud" GKE features like HPA/VPA/Autopilot

> 
> @abeatrix lmk if you disagree w/ any of the above.

rafax commented:
> We have built a kustomize component that would update the env vars in frontend automatically when deploying using kustomize; however, if service discovery is preferred, the site admins / cluster admins do have the option to add service discovery to frontend anytime (using kustomize).

If we have an option to support service discovery (and therefore, can scale deployments without reconfiguring clients / redeploying customize) we're good on Cloud side.

keegancsmith commented:
@abeatrix I'll take a look tomorrow, but I noticed that there are a lot of conflicts. You probably wanna merge in origin/main or rebase onto it.

abeatrix commented:
adding @slimsag as a reviewer to make sure the new changes will also work with App as expected

slimsag commented on CHANGELOG.md:
```diff
@@ -41,6 +42,7 @@ All notable changes to Sourcegraph are documented in this file.
 
 - Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
 - Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)
+- Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)
```
intentionally added?

abeatrix replied:
Nope! Removed! Thanks for catching that!

slimsag commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -494,22 +496,42 @@ var (
+	deployType, ok := getEnv(environ, deploy.DeployTypeEnvName)
+	if !ok {
+		deployType = deploy.Default
+	}
```
I think using `deploy.Type()` here is better, it has some better handling logic + allows for mocking: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7027cb696a8cdfc9eae3d8e01e8296d26392ac94/-/blob/internal/conf/deploy/deploytype.go?L21-35

abeatrix replied:
Updated to use deploy.Type() and deploy.Mock() for testing

slimsag commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +606,149 @@ var (
+	deployType, ok := getEnv(environ, deploy.DeployTypeEnvName)
+	if !ok {
+		deployType = deploy.Default
+	}
```
same here (use `deploy.Type()`)

abeatrix replied:
Updated to use deploy.Type() and deploy.Mock() for testing

slimsag commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +606,149 @@ var (
+	deployType, ok := getEnv(environ, deploy.DeployTypeEnvName)
+	if !ok {
+		deployType = deploy.Default
+	}
```
same here

abeatrix replied:
Updated to use deploy.Type() and deploy.Mock() for testing

slimsag commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +606,149 @@ var (
+	deployType, ok := getEnv(environ, deploy.DeployTypeEnvName)
+	if !ok {
+		deployType = deploy.Default
+	}
```
same here

abeatrix replied:
Updated to use deploy.Type() and deploy.Mock() for testing

slimsag commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +606,149 @@ var (
+	case deploy.DockerCompose:
+		fmtStrTail = fmt.Sprintf(":%s", port)
+	default:
+		return "", errors.New("Error: unsupported deployment type: " + deployType)
```
I think this would be problematic in a few environments where the env var is set explicitly. Basically you just need to ensure we are handling all the deploy types here, the missing ones are:

```
	SingleDocker  = "docker-container"
	PureDocker    = "pure-docker"
	Dev           = "dev"
	SingleProgram = "single-program"
```

slimsag replied:
I thought you'd need to handle these but it seems to work fine in single-program mode 🤔 I'm not sure, maybe you don't need to handle the other cases here and I am missing something about how this code works when the env var is set explicitly

abeatrix replied:
@slimsag We are only adding support for docker-compose and kubernetes environments atm. For those environments, we will just return what they have input in https://sourcegraph.com/github.com/sourcegraph/sourcegraph@99af7c30e464f3be5e1a5dfd165ef334c10b2616/-/blob/cmd/frontend/internal/cli/config.go?L726-729 if this makes sense?

slimsag commented on internal/conf/deploy/deploytype.go:
```diff
@@ -11,9 +11,18 @@ const (
+
+// DeployTypeEnvName is the environment variable name that users can use to
+// specify the deployment type.
+const DeployTypeEnvName = "DEPLOY_TYPE"
```
I'd revert this addition, if someone relies on the env var name they will be missing handling for ldflags which `deploy.Type` accounts for.

abeatrix replied:
Reverted

slimsag reviewed (approved):
LGTM aside from my comments

keegancsmith commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +601,136 @@ var (
+		return addrs, err
+	}
+
+	// Not set, use the default (non-service discovery on searcher)
```
```suggestion
	// Not set, use the default (non-service discovery on symbols)
```

keegancsmith commented on cmd/frontend/internal/cli/config_test.go:
```diff
@@ -165,11 +167,181 @@ func TestReadSiteConfigFile(t *testing.T) {
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			deploy.Mock(tc.deployType)
+			got, _ := gitserverAddr(tc.environ)
```
we should assert on errors as well. A reasonable pattern here is just checking if err contains a subset of wantErr. Sketch code:
```suggestion
			got, err := gitserverAddr(tc.environ)
			if (tc.wantErr == "") != (err == nil) {
				t.Errorf("unexpected error. want=%q got=%v", tc.wantErr, err)
			} else if err != nil && !strings.Contains(err.Error(), tc.wantErr) {
				t.Errorf("error does not contain %q: %v", tc.wantErr, err)
			}
```

keegancsmith commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +601,136 @@ var (
+	case deploy.DockerCompose:
+		fmtStrTail = fmt.Sprintf(":%s", port)
+	default:
+		return "", errors.New("Error: unsupported deployment type: " + deployType)
```
nit: errors in go convenetionally start lowercase (to aid readability in wrapping)
```suggestion
		return "", errors.Errorf("unsupported deployment type: %s", deployType)
```

keegancsmith commented on internal/symbols/client.go:
```diff
@@ -32,10 +32,23 @@ import (
+			return conns.Symbols
+		})
+	})
+	return symbolsURLs.String()
```
Blocker: This change is not safe. It will use a static configuration that won't change over the lifetime of the process. Additionally I worry about potential deadlocks if the conf package is not ready given this is called inside of LoadConfig.

keegancsmith replied:
This required a bit more context to fix, so I went ahead and pushed a fix for this :)

keegancsmith reviewed (changes requested):
Nice code!

One blocking comment inline, so requesting changes.

abeatrix commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -574,48 +601,136 @@ var (
+	symbolsURLsOnce.Do(func() {
+		addr, err := symbolsAddr(os.Environ())
+		if err != nil {
+			symbolsURLs = endpoint.Empty(errors.New("a symbols service has not been configured"))
```
wrapped the rest of the errs :)
